### PR TITLE
RD-4796 Add user configuration files snapshot endpoints

### PR DESCRIPTION
**Pull request by kubama**
## Description
Added user configuration files snapshot endpoints.
Broke down the snapshot endpoints spec into parts due to emerging maintenance issues.

## Screenshots / Videos
N/A

## Checklist

- [X] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [X] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [X] I added proper labels to this PR.

## Tests
Added unit tests

 

## Documentation
N/A

### CODE REVIEW

File: backend/routes/Snapshots.ts [hunk 1]
```diff
@@ -11,11 +12,16 @@ import { checkPageGroupExists, createPageGroup, getUserPageGroups } from '../han
...
 import type { BlueprintUserDataInstance } from '../db/models/BlueprintUserDataModel';
 import type { BlueprintUserDataAttributes } from '../db/models/BlueprintUserDataModel.types';
+import { getResourcePath } from '../utils';
 
 const router = express.Router();
 
 router.use(express.json());
 
+function sendConflictResponse(res: Response) {
+    res.status(400).send({ message: 'Snapshot data conflicts with already existing data' });
```
Comment (PatrycjaKubiczek) [inline backend/routes/Snapshots.ts:22]: shouldn't the message be part of `en.json`? 🤔

Comment (kubama) [reply backend/routes/Snapshots.ts:22]: We don't have i18n in the backend

**Review by PatrycjaKubiczek — DISMISSED** [review]

File: backend/routes/Snapshots.ts [hunk 2]
```diff
@@ -165,11 +170,45 @@ export type PageGroupsSnapshot = Omit<PageGroup, CommonPropertiesToOmit>[];
...
+        res.send(snapshot);
+    });
+
+    router.post<never, { message: string }, ConfigurationSnapshot>('/configuration', (req, res, next) => {
+        if (union(userConfigurationFiles, Object.keys(req.body)).length > userConfigurationFiles.length) {
+            res.status(400).send({ message: 'Unsupported configuration file' });
+            return;
+        }
+
+        const existingFile = Object.keys(req.body).find(fileName => existsSync(getResourcePath(fileName, true)));
```
Comment (qooban) [inline backend/routes/Snapshots.ts:201]: Optional:

Maybe we can extract `Object.keys(req.body)` to a constant and reuse.

File: backend/test/routes/snapshots/configuration.test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    it('restores snapshot data', () => {
+        return request(app)
+            .post('/console/snapshots/configuration')
+            .send({ 'overrides.json': {} })
+            .then(response => {
+                expect(response.statusCode).toBe(201);
+                expect(writeJson).toHaveBeenCalledWith(expect.stringMatching('/userData/overrides.json$'), {});
+            });
+    });
```
Comment (qooban) [inline backend/test/routes/snapshots/configuration.test.ts:25]: These tests will fail if we have one of the configuration files defined in `/userData` folder.
What do you think about introducing similar solution to what we have in https://github.com/cloudify-cosmo/cloudify-stage/blob/366fb70f6f9043af89cfbc2d3969d54f68278778/backend/test/migration.test.ts for backing it up?

Comment (kubama) [reply backend/test/routes/snapshots/configuration.test.ts:25]: I think they should be working fine due to mocked `fs-extra`

File: backend/routes/Snapshots.ts (hunk 2 above)
Comment (kubama) [reply backend/routes/Snapshots.ts]: Done

File: backend/test/routes/snapshots/configuration.test.ts (hunk 3 above)
Comment (qooban) [reply backend/test/routes/snapshots/configuration.test.ts:25]: Right. Didn't notice it, all good then.

**Review by qooban — APPROVED** [review]

**Pull request merged**